### ISSUE TITLE
BadLocationException when converting the undo history from Unicode to ASCII

### ISSUE THREAD

**Issue by lemmy** [tlaplus/tlaplus#72]
1. Activate "Use Unicode"
2. Open e.g. switcherroo.tla and add and remove a couple of newlines before end of module marker
3. Save spec editor
4. Deactivate "Use Unicode"

@pron What is the undo history being converted for?

**Comment by lemmy** [comment]
@pron It seems the code removes the Unicode <> ASCII conversion operation from the history?!

**Comment by pron** [comment]
> It seems the code removes the Unicode <> ASCII conversion operation from the history?!

Yes (as does the code that adds those two versioning lines at the bottom). The reason is that the undo history is local, while the Unicode mode is global. If we were to allow this, the buffer would revert to its former state without changing the mode.

> What is the undo history being converted for?

Sigh. This is one of the trickiest hacks, and I believe I reported this as a known bug when I submitted the change. Basically, the goal is to support undo operations even after a mode change, but it's generally very hard to do, so it's a "best effort". IIRC, the exceptions are harmless, but the undo operations may not behave well.

**Comment by lemmy** [comment]
What about purging the undo history when switching between modes?

**Comment by pron** [comment]
Sure, that could work, too, if that's acceptable.

**Comment by lemmy** [comment]
Can you think of a use case where one switches between Unicode and ASCII mode back and forth?

**Comment by pron** [comment]
Probably not, but people do all kinds of things...

**Comment by lemmy** [comment]
https://github.com/tlaplus/tlaplus/blob/33b11f8c70b1bb7cb3770287272666f51f970e7a/org.lamport.tla.toolbox.editor.basic/src/org/lamport/tla/toolbox/editor/basic/TLAEditor.java#L337-L344

**Comment by lemmy** [comment]
See branch https://github.com/tlaplus/tlaplus/tree/pron-unicode
